bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java

```diff
@@ -0,0 +1,278 @@
+        super();
+
+        if (expression == null) {
+            throw new IllegalArgumentException("The expression cannot be null");
+        }
+
+        this.expression = expression;
+        this.delimiters = delimiters;
+        this.startDate = startDate;
+        this.timeZone = timeZone;
+        this.minimumCandidates = minimumCandidates;
+
+        setStartDate(startDate);
+        setTimeZone(timeZone);
+        parseExpression(expression);
+    }
+
+    @Override
+    public final Date getStartDate() {
+        if (startDate == null) {
+            startDate = Calendar.getInstance(getTimeZone()).getTime();
+        }
+        return startDate;
+    }
+
+    @Override
+    public void setStartDate(Date startDate) throws IllegalArgumentException, ParseException {
+        if (startDate == null) {
+            throw new IllegalArgumentException("The start date of the rule must not be null");
+        }
+        this.startDate = startDate;
+        parseExpression(expression);
+    }
+
+    @Override
+    public TimeZone getTimeZone() {
+        if (timeZone == null) {
+            timeZone = TimeZone.getDefault();
+        }
+
+        return timeZone;
+    }
+
+    @Override
+    public void setTimeZone(TimeZone timeZone) throws IllegalArgumentException, ParseException {
+        if (timeZone == null) {
+            throw new IllegalArgumentException("The time zone must not be null");
+        }
+        this.timeZone = timeZone;
+        parseExpression(expression);
+    }
+
+    @Override
+    public String getExpression() {
+        return expression;
+    }
+
+    @Override
+    public void setExpression(String expression) throws ParseException {
+        this.expression = expression;
+        parseExpression(expression);
+    }
+
+    @Override
+    public String toString() {
+        return expression;
+    }
+
+    /**
+     * Parse the given expression
+     *
+     * @param expression - the expression to parse
+     * @throws ParseException - when the expression cannot be successfully be parsed
+     * @throws IllegalArgumentException - when expression parts conflict with each other
+     */
+    public void parseExpression(String expression) throws ParseException, IllegalArgumentException {
+
+        StringTokenizer expressionTokenizer = new StringTokenizer(expression, delimiters, false);
+        int position = 0;
+
+        setExpressionParts(new ArrayList<E>());
+        setCandidates(new ArrayList<Date>());
+
+        while (expressionTokenizer.hasMoreTokens()) {
+            String token = expressionTokenizer.nextToken().trim();
+            position++;
+            getExpressionParts().add(parseToken(token, position));
+        }
+
+        Collections.sort(getExpressionParts());
+
+        validateExpression();
+
+        if (startDate == null) {
+            setStartDate(Calendar.getInstance().getTime());
+        }
+
+        applyExpressionParts();
+
+        synchronized (this) {
```

Comment:
You are using a synchronized block here, all fine.
In the applyExpressionsPart function you iterate over the expression parts without using some sync-mechanism that setExpressionsPart could be called.
As this is the only synchronized block are you sure that using this will make the objects of this classes thread safe?